becquerel/core/spectrum.py

```diff
@@ -456,6 +461,62 @@ def from_file(cls, infilename):
 
         return cls(**kwargs)
 
+    @classmethod
+    def from_listmode(cls, listmode_data, timestamps=None, bins=None,
+                      xmin=None, xmax=None):
+        """Construct a Spectrum object from an array of (possibly timestamped)
+        listmode data.
+
+        Args:
+          listmode_data: the array containing the listmode data, e.g., the ADC
+            channels triggered or pulse integral values recorded
+          bins: integer number of bins OR array of bin_edges
+          xmin: minimum x of histogram; equals bin_edges[0]
+          xmax: maximum x of histogram; equals bin_edges[-1] + bin_widths[-1]
+          timestamps: array of timestamps associated with listmode_data
+
+        Returns:
+          A Spectrum object
+
+        Raises:
+          AssertionError: no listmode_data, or xmin >= xmax, or nbins < 1
+        """
+
+        assert len(listmode_data) > 1
+
+        if xmin is None:
+            xmin = 0
+        if xmax is None:
+            xmax = np.ceil(max(listmode_data))
+        if bins is None:
+            bins = np.arange(xmin, xmax + 1, dtype=np.int)
+
+        assert xmin < xmax
+        if isinstance(bins, int):
+            assert bins > 0
+        else:
+            assert len(bins) > 1
+
+        bin_counts, bin_edges = np.histogram(listmode_data,
+                                             bins=bins,
+                                             range=(xmin, xmax))
+        kwargs = {'counts': bin_counts,
+                  'bin_edges_kev': bin_edges}  # TODO: introduce bin_edges_adc
+
+        if timestamps is not None:
+            assert len(timestamps) == len(listmode_data)
+            # TODO: How do we handle these? This isn't accurate ...
+            # What are the timestamps: 'float', 'int',
+            #   'datetime', epoch [nano][micro][milli]seconds?
+            warnings.warn('Using first/last timestamps to approximate ' +
+                          'start / stop_time and realtime.', SpectrumWarning)
+            kwargs['start_time'] = min(timestamps)
+            kwargs['stop_time'] = max(timestamps)
+            kwargs['realtime'] = kwargs['stop_time'] - kwargs['start_time']
+            kwargs['livetime'] = None
```

Comment:
I think it would be more intuitive if instead of setting those from a timestamp array that can change from use case to use case, we just pass all the unused kwargs to the class creator. So if someone cares about `livetime`, he can set it the way he thinks it needs to be set, i.e. `livetime=timstamps[-1] - timestamps[0]`, (same for `start_time`, `stop_time` and `realtime`, which are never really used anyways). We can not impossible cover all cases that might arise here, better let the user make a wise decision.

Reply:
Sure, makes sense. This also saves us from converting back and forth between `datetime` and `float` etc depending on timestamps vs time durations like `livetime` and `realtime`, and saves us from always approximating the start/stop acquisition times as the first/last timestamp.

Since the `kwargs` are already passed to the class creator, this just becomes:

    ...
    if timestamps is not None:
            assert len(timestamps) == len(listmode_data)
    
    return cls(**kwargs)

Reply:
I would not have a timestamps field at all as it won't be used. Something like

```
    def from_listmode(cls, listmode_data, timestamps=None, bins=None,
                      xmin=None, xmax=None, **kwargs):
        ...
        kwargs['counts'] =  bin_counts
        kwargs['bin_edges_kev' = bin_edges
        ...
        return cls(**kwargs)
```

Reply:
Fine by me. I've removed them in the latest commit. The user can always do `spec.timestamps = ...` if necessary.